File: packages/core/src/trace/DdTrace.ts
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

import { InternalLog } from '../InternalLog';
import { SdkVerbosity } from '../config/types/SdkVerbosity';
import type { DdNativeTraceType } from '../nativeModulesTypes';
import { encodeAttributes } from '../sdk/AttributesEncoding/attributesEncoding';
import {
    bufferNativeCallReturningId,
    bufferNativeCallWithId
} from '../sdk/DatadogProvider/Buffer/bufferNativeCall';
import { getNativeDdTrace } from '../specs/NativeDdTrace';
import type { DdTraceType } from '../types';
import { getGlobalInstance } from '../utils/singletonUtils';
import { DefaultTimeProvider } from '../utils/time-provider/DefaultTimeProvider';

const TRACE_MODULE = 'com.datadog.reactnative.trace';

const timeProvider = new DefaultTimeProvider();

class DdTraceWrapper implements DdTraceType {
    private nativeTrace: DdNativeTraceType = getNativeDdTrace() as DdNativeTraceType;

    startSpan = (
        operation: string,
        context: object = {},
        timestampMs: number = timeProvider.now()
    ): Promise<string> => {
        const spanId = bufferNativeCallReturningId(() =>
            this.nativeTrace.startSpan(
                operation,
                encodeAttributes(context),
                timestampMs
            )
        );
        InternalLog.log(`Starting span “${operation}”`, SdkVerbosity.DEBUG);
        return spanId;
    };

    finishSpan = (
        spanId: string,
        context: object = {},
        timestampMs: number = timeProvider.now()
    ): Promise<void> => {
        InternalLog.log(`Finishing span #${spanId}`, SdkVerbosity.DEBUG);
        return bufferNativeCallWithId(
            id =>
                this.nativeTrace.finishSpan(
                    id,
                    encodeAttributes(context),
                    timestampMs
                ),
            spanId
        );
    };
}

export const DdTrace: DdTraceType = getGlobalInstance(
    TRACE_MODULE,
    () => new DdTraceWrapper()
);

File: packages/core/android/src/newarch/kotlin/com/datadog/reactnative/DdTrace.kt
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

package com.datadog.reactnative

import com.facebook.fbreact.specs.NativeDdTraceSpec
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap

/**
 * The entry point to use Datadog's Trace feature.
 */
class DdTrace(
    reactContext: ReactApplicationContext,
) : NativeDdTraceSpec(reactContext) {

    private val implementation = DdTraceImplementation()

    override fun getName(): String = DdTraceImplementation.NAME

    /**
     * Start a span, and returns a unique identifier for the span.
     * @param operation The operation name of the span.
     * @param context The additional context to send.
     * @param timestampMs The timestamp when the operation started (in milliseconds).
     * If not provided, current timestamp will be used.
     */
    @ReactMethod
    override fun startSpan(operation: String, context: ReadableMap, timestampMs: Double, promise: Promise) {
        implementation.startSpan(operation, context, timestampMs, promise)
    }

    /**
     * Finish a started span.
     * @param spanId The unique identifier of the span.
     * @param context The additional context to send.
     * @param timestampMs The timestamp when the operation stopped (in milliseconds).
     * If not provided, current timestamp will be used.
     */
    @ReactMethod
    override fun finishSpan(spanId: String, context: ReadableMap, timestampMs: Double, promise: Promise) {
        implementation.finishSpan(spanId, context, timestampMs, promise)
    }
}
